ykshim_client/src/lib.rs

```diff
@@ -68,15 +67,15 @@ pub struct ThreadTracer(*mut RawThreadTracer);
 /// Start tracing using the specified kind of tracing.
 pub fn start_tracing(tracing_kind: TracingKind) -> ThreadTracer {
     let tracer = unsafe { __ykshim_start_tracing(tracing_kind as u8) };
-    assert!(!tracer.is_null());
+    debug_assert!(!tracer.is_null());
     ThreadTracer(tracer)
 }
 
 impl ThreadTracer {
     pub fn stop_tracing(mut self) -> Result<SirTrace, CString> {
         let mut err_msg = std::ptr::null_mut();
         let sir_trace = unsafe { __ykshim_stop_tracing(self.0, &mut err_msg) };
```

Comment:
Ah! I think what you really want here is to use `mem::replace` which prevents the potential race that you're currently using two lines to avoid.

Reply:
I'm not sure what you mean. There's no race that I'm aware of.

This dance is because the use of raw pointers across the C ABI kills ownership handling and we have to track it ourself:

 - If `self` is dropped without a shim function consuming it, then the client side must drop it to avoid a leak.
 - If `self` is consumed by a shim, then the other side is responsible for dropping it. If the client side also drops it we have a double free.

The null pointer is used to mark `self` consumed, and the check is [here](https://github.com/softdevteam/yk/pull/260/files/8a2c3f62e767a1020d2a10edd2ceaa2559111acc..c9bd3ce8a103045a3aa03882887143d1a5267f3e#diff-01f309b981ab8e667ec6d6662d6d1c8cb17071b932bdb1748dfd9cf869236fdeR88).

(We only use this idiom for types where there exists a shim which consumes)

I'm unsure how `mem::replace` helps, but I may have missed something?

Reply:
There isn't a race here, because you carefully:

1. read the value (line 77).
2. now that it's stale, zero it (line 78).

If you didn't do that second step, you'd have a race.

What I'm suggesting is that I *think* you're doing an explicit version of `mem::replace` and it might be worth using it instead of the two steps above? Something like:

```rust
let v = mem::replace(self.0, ptr::null_mut());
let sir_trace = unsafe { __ykshim_stop_tracing(v, &mut err_msg) };
```

Reply:
Is there any worth in moving a raw pointer though? It's already `Copy`.

I think I'm missing something...

Reply:
I think it's worth considering `replace` here because it makes clear that you can't accidentally have a stale value lurking around. Up to you.

Reply:
Resolved offline.